libutils/glob_lib.c

```diff
@@ -0,0 +1,446 @@
+ * Returns a sequence of substrings split on delimiter.
+ */
+static Seq *SplitString(const char *const str, const char *const del)
+{
+    assert(str != NULL);
+    assert(del != NULL);
+
+    Seq *const seq = SeqNew(1, free);
+
+    const char *start = str;
+    const char *end = strpbrk(str, del);
+
+    while (end != NULL)
+    {
+        char *const tmp = xstrndup(start, end - start);
+        SeqAppend(seq, tmp);
+        start = end + 1;
+        end = strpbrk(start, del);
+    }
+
+    char *const tmp = xstrdup(start);
+    SeqAppend(seq, tmp);
+
+    return seq;
+}
+
+/**
+ * Returns a set of brace-expanded patterns.
+ */
+static void ExpandBraces(const char *pattern, StringSet *expanded)
+    FUNC_UNUSED; // Unused if compiled without PCRE2, but tested in unit tests
+
+static void ExpandBraces(const char *const pattern, StringSet *const expanded)
+{
+    assert(pattern != NULL);
+    assert(expanded != NULL);
+
+    char *const start = xstrdup(pattern);
+
+    /* First we will split the pattern into three parts based on the curly
+     * brace delimiters. We look for the shortest possible match. E.g., the
+     * string "foo{{bar,baz}qux,}" becomes "foo{", "bar,baz", "qux,}".
+     */
+    char *left = NULL, *right = NULL;
+    for (char *ch = start; *ch != '\0'; ch++)
+    {
+        if (*ch == '{')
+        {
+            left = ch;
+        }
+        else if (left != NULL && *ch == '}')
+        {
+            right = ch;
+            break;
+        }
+    }
+
+    /* Next we check if base case is reached. I.e., if there is no curly brace
+     * pair to expand, then add the pattern itself to the result.
+     */
+    if (right == NULL)
+    {
+        StringSetAdd(expanded, start);
+        return;
+    }
+    *left = '\0';
+    *right = '\0';
+
+    /* Next we split the middle part (between the braces) on the comma
+     * delimiter. E.g., the string "bar,baz" becomes "bar", "baz".
+     */
+    char *middle = left + 1, *end = right + 1;
+    Seq *const split = SplitString(middle, ",");
+
+    /* Lastly we combine the three parts for each split element, and
+     * recursively expand subsequent curly braces.
+     */
+    const size_t len = SeqLength(split);
+    for (size_t i = 0; i < len; i++)
+    {
+        middle = SeqAt(split, i);
+        char *const next = StringConcatenate(3, start, middle, end);
+        ExpandBraces(next, expanded);
+        free(next);
+    }
+
+    free(start);
+    SeqDestroy(split);
+}
+
+#ifdef WITH_PCRE2
+
+/**
+ * Translate the square bracket part of a shell expression.
+ */
+static size_t TranslateBracket(
+    const char *const pattern, const size_t n, size_t i, Buffer *const buf)
+{
+    size_t j = i;
+    if (j < n && pattern[j] == '!')
+    {
+        j += 1;
+    }
+    if (j < n && pattern[j] == ']')
+    {
+        j += 1;
+    }
+    while (j < n && pattern[j] != ']')
+    {
+        j += 1;
+    }
+    if (j >= n)
+    {
+        BufferAppendString(buf, "\\[");
+    }
+    else
+    {
```

Comment:
All of this seems quite cryptic to me, could you add some comments or asserts or more descriptive variable names?

Reply:
Yes, same here. It is very cryptic and hard to explain with words. Tried to use better variable names, but not sure if it is that much better.